src/Pools/Pool.php

```diff
@@ -106,13 +107,29 @@ public function setReconnectSleep(int $reconnectSleep): self
     }
 
     /**
-     * @return self
+     * Summary: 
+     *  1. Try to get a connection from the pool
+     *  2. If no connection is available, wait for one to be released
+     *  3. If still no connection is available, throw an exception
+     *  4. If a connection is available, return it
+     * 
+     * @return Connection
      */
-    public function fill(): self
+    public function pop(): Connection
     {
-        $this->pool = [];
+        $connection = array_pop($this->pool);
+
+        if(is_null($connection)) { // pool is empty, wait an if still empty throw exception
+            usleep(50000); // 50ms TODO: make this configurable
 
-        for ($i=0; $i < $this->size; $i++) { 
+            $connection = array_pop($this->pool);
+
+            if(is_null($connection)) {
+                throw new Exception('Pool is empty');
+            }
+        }
```

Comment:
This is essentially checking twice in the worst case. Is that all we want ? Or should it be a loop with where we check if max attempts have been reached?

Reply:
I think we need something more sophisticated, but was planning to leave it for later.